library/x509_create.c

```diff
@@ -86,73 +92,208 @@ static const char *x509_at_oid_from_name( const char *name, size_t name_len )
     return( cur->oid );
 }
 
+static int x509_dn_check_special_chars( const char **cur, const char *end )
+{
+    const char *c = *cur;
+    int ret = 0;
+
+    /* Check for valid escaped characters */
+    if( c == end || *c != ',' )
+    {
+        ret = MBEDTLS_ERR_X509_INVALID_NAME;
+    }
+
+    return( ret );
+}
+
 int mbedtls_x509_string_to_names( mbedtls_asn1_named_data **head, const char *name )
 {
     int ret = 0;
-    const char *s = name, *c = s;
-    const char *end = s + strlen( s );
+    /*
+     * `start` is a pointer into the 'name' buffer.
+     * Depending on the parsing state, it either points to the beginning of
+     * the currently parsed DN OID tag or the currently parsed DN value.
+     * `cur` points to the current character in the string being parsed.
+     * `d` acts as a write pointer into the temporary 'data' buffer which holds
+     * the unescaped and unquoted DN values found in 'name'.
+     */
+    const char *start = name, *cur = start;
+    const char *end = start + strlen( start );
     const char *oid = NULL;
-    int in_tag = 1;
     char data[MBEDTLS_X509_MAX_DN_NAME_SIZE];
     char *d = data;
+    x509_dn_parse_state dn_cur_state = dn_in_tag;
+    size_t tag_len = 0;
 
     /* Clear existing chain if present */
     mbedtls_asn1_free_named_data_list( head );
 
-    while( c <= end )
+    while( cur <= end )
     {
-        if( in_tag && *c == '=' )
+        switch( dn_cur_state )
         {
-            if( ( oid = x509_at_oid_from_name( s, c - s ) ) == NULL )
+            /*
+             * Currently parsing a DN tag.
+             */
+            case dn_in_tag:
             {
-                ret = MBEDTLS_ERR_X509_UNKNOWN_OID;
-                goto exit;
+                if( *cur == '=' )
+                {
+                    if( ( oid = x509_at_oid_from_name( start, tag_len ) ) == NULL )
```

Comment:
As mentioned above, we're still not handling whitespaces prior to `=` correctly, e.g. `A =B` wouldn't be parsed correctly as far as I see.

Reply:
@RonEld Sorry, I missed that. A whitespace terminates the current tag, though, while the present code keeps parsing; for example, `A BC=D` would be parsed as the OID `A B` currently.

Reply:
I have also added a [test case](https://github.com/ARMmbed/mbedtls/pull/784/files#diff-0807f152469bce0d302ec7282849e58bR131) for this

Reply:
I mean, whitespaces in the middle of a tag is not permitted

Reply:
@RonEld Yes, but we're currently not checking for that, or do we?

Reply:
`x509_at_oid_from_name()` will return an error in this case, as returned in the test"X509 String to Names #19"

Reply:
I think the current code wouldn't handle ` A=B` correctly, because it'd parse the OID as ` ` instead of `A`.

Reply:
you may have a point. I'll add a test case for this and fix if needed